posthog/api/query.py

```diff
@@ -222,57 +243,112 @@ async def query_awaited(request: Request, *args, **kwargs) -> StreamingHttpRespo
-                await asyncio.sleep(FAST_POLL_INTERVAL)
-            elif elapsed_time < MEDIUM_POLL_DURATION:
-                await asyncio.sleep(MEDIUM_POLL_INTERVAL)
-            else:
-                await asyncio.sleep(SLOW_POLL_INTERVAL)
-
-    return StreamingHttpResponse(
-        event_stream(),
-        content_type="text/event-stream",
-        headers={
-            "Cache-Control": "no-cache",
-            "X-Accel-Buffering": "no",
-        },
-    )
+                while time.time() - start_time < MAX_QUERY_TIMEOUT:
+                    # Check if the query task has completed
+                    if query_task.done():
+                        try:
+                            result = query_task.result()
+                        except (ExposedHogQLError, ExposedCHQueryError) as e:
+                            yield f"data: {json.dumps({'error': str(e), 'status_code': 400})}\n\n".encode()
+                        except Exception:
+                            yield f"data: {json.dumps({'error': 'Server error'})}\n\n".encode()
+
+                        if isinstance(result, BaseModel):
+                            yield f"data: {result.model_dump_json(by_alias=True)}\n\n".encode()
+                        else:
+                            yield f"data: {json.dumps(result)}\n\n".encode()
+                        break
+
+                    try:
+                        # Try to get a status updates while waiting
+                        current_time = time.time()
+                        if current_time - last_update_time >= UPDATE_INTERVAL:
+                            status = await sync_to_async(manager.get_clickhouse_progresses)()
+
+                            if isinstance(status, BaseModel):
+                                yield f"data: {status.model_dump_json(by_alias=True)}\n\n".encode()
+                                last_update_time = current_time
+                    # Just ignore errors when getting progress, shouldn't impact users
+                    except Exception as e:
+                        capture_exception(e)
+
+                    elapsed_time = time.time() - start_time
+                    if elapsed_time < FAST_POLL_DURATION:
+                        await asyncio.sleep(FAST_POLL_INTERVAL)
+                    elif elapsed_time < MEDIUM_POLL_DURATION:
+                        await asyncio.sleep(MEDIUM_POLL_INTERVAL)
+                    else:
+                        await asyncio.sleep(SLOW_POLL_INTERVAL)
+
+            finally:
+                # If we break the loop early, ensure we cancel the query task
+                if not query_task.done():
+                    query_task.cancel()
+                    yield f"data: {json.dumps({'error': 'Query cancelled'})}\n\n".encode()
+
+        return StreamingHttpResponse(
+            event_stream(),
+            content_type="text/event-stream",
+            headers={
+                "Cache-Control": "no-cache",
```

Comment:
```suggestion
                "Cache-Control": "no-cache",
                "Connection": "keep-alive,
```
Just looking at livestream's config for ES you _might_ want to set the `Connection=keep-alive` header
https://github.com/PostHog/posthog/blob/master/livestream/served.go#L116